proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,135 @@
+  repeated google.protobuf.ListValue expected_products = 4;
+
+  // This field contains the expected command used to perform the step.
+  repeated string expected_command = 5;
+
+  // This field contains a value that indicates the minimum number of keys that
+  // need to be used to sign the step's in-toto attestation.
+  int64 threshold = 6;
+}
+
+// This defines the format used to record keys used in the software supply
+// chain. An in-toto link is attested using one or more keys defined in the
+// in-toto layout.
+message SigningKey {
+  // key_id is an identifier for the signing key.
+  string key_id = 1;
+
+  // This field identifies the hash algorithms used in the particular key ID.
+  repeated string key_id_hash_algorithms = 2;
+
+  // This field identifies the specific signing method. Eg: "RSASSA-PSS",
+  // "ed25519", and "ecdsa".
+  string key_type = 3;
+
+  // This field contains the actual public key.
+  string public_key_value = 4;
+
+  // This field contains the overall scheme used for the creation of the signing
+  // key. Eg: "rsassa-pss-sha256".
+  string key_scheme = 5;
+}
+
+// This corresponds to a signed in-toto attestation - it is made up of one or
+// more signatures and the link attestation itself. This is used for occurrences
+// of a Grafeas in-toto note.
+message Details {
+  repeated Signature signatures = 1;
+  Link link = 2 [json_name="signed"];
+}
+
+// A signature object consists of the KeyID used and the signature itself.
+message Signature {
+  string key_id = 1 [json_name="keyid"];
+  string signature = 2 [json_name="sig"];
+}
+
+// This corresponds to an in-toto link attestation.
+message Link {
+  // This field contains the full command executed for the step. This can also
+  // be empty if attestations are generated for operations that aren't directly
+  // mapped to a specific command. Each term in the command is an independent
+  // string in the list. An example of a command in the in-toto metadata field
+  // is:
+  // "command": ["git", "clone", "https://github.com/in-toto/demo-project.git"]
+  repeated string effective_command = 1 [json_name="command"];
+
+  // Materials are the supply chain artifacts that go into the step and are used
+  // for the operation performed. The key of the map is the path of the artifact
+  // and the structure contains the recorded hash information. An example is:
+  // "materials": {
+  //   "foo/bar": {
+  //     "sha256": "ebebf..."
+  //   }
+  // }
+  map<string, google.protobuf.Struct> materials = 2;
+
+  // Products are the supply chain artifacts generated as a result of the step.
+  // The structure is identical to that of materials.
+  map<string, google.protobuf.Struct> products = 3;
+
+  // By Products are data generated as part of a software supply chain step, but
+  // are not the actual result of the step. The suggested fields are "stderr",
+  // "stdout", and "return-value".
+  map<string, google.protobuf.Struct> byproducts = 4;
```

Comment:
ditto here for defining a separate type

In general, defining specific message types is a very common practice in proto design.

Reply:
`byproducts` and `environment` have no strict structure - they can be empty - and their contents are not verified by in-toto. For example, here's an example in-toto link that Datadog is using: [tag.c295cf63.link](https://dd-integrations-core-wheels-build-stable.datadoghq.com/targets/in-toto-metadata/3e592d47ab815cb9bd8aa064bbaa2a217c1baa9a55a7f8b73d2da6dfd14a247e/tag.c295cf63.link). They don't populate either field.

Reply:
Hm, from the description above, the structure is `artifact: values`, where `artifact` is a binary/pkg/container in Grafeas parlance, so is represented by `cpe_uri` -- am I misunderstanding the description in the comment? If that's the case, how about making just a generic

```
message Material {
  map<string, string> artifact_to_materials = 1;
}
```
?

Also, PTAL at the proto design resources I shared, it might be helpful in clarifying what I'm asking here.:)

Reply:
I think we may be going around in circles a bit here, and I may be misunderstanding your suggestion. So let me summarise for my own clarity. The `materials` and `products` field in link metadata are maps where each key is an artifact identifier (currently file paths), and each value is another map. The keys in this inner map are hash algorithms, and the values are the hashes of the artifact using the corresponding hash algorithm.
```
"materials":  {
  "<artifact>: {
    "sha256": "<hash>",
    "sha512": "<hash>",
  }
}
```
What I'm really trying to define using protobuf is `map<string, map<string, string>> materials = 2;`. When I found I couldn't do that, the closest I found in the docs was [google.protobuf.String](https://developers.google.com/protocol-buffers/docs/proto3#json) in place of the inner map. I'd like to make this specific, and ensure that the inner map also just contains `string: string` but I haven't found a way so far. I did try using a message as follows:
```
  message ArtifactHashes {
    map<string, string> hashes = 1;
  }

  map<string, ArtifactHashes> materials = 2;
```
This has the effect of requiring `hashes` as a key in the JSON we POST to the server, which differs from the in-toto spec. Example:
```
          "materials": {},
          "products": {
            "demo-project/foo.py": {
              "hashes": {
                "sha256": "ebebf8778035e0e842a4f1aeb92a601be8ea8e621195f3b972316c60c9e12235"
              }
            }
          },
```

Reply:
Is the list of supported hash algorithms known? If so, we could define the `ArtifactHashes` as:
```
  message ArtifactHashes {
    string sha256 = 1;
    string sha512 = 2;
  }
```

Reply:
The specification does not enforce any specific hash algorithms. The in-toto reference implementation currently implements SHA-256. I'm okay to use this definition and we can approach this further down the line, as and when the algorithms need to be updated. :)

Reply:
Added.